### PR TITLE
KAFKA-7772: Adjust log levels of classes in Connect via JMX

### PR DESCRIPTION
**Pull request by wicknicks**
Consolidating ways to adjust log levels in AK. This change provides a common way to modify log levels of running Kafka brokers and Kafka Connect workers via a JMX interface.

Signed-off-by: Arjun Satish <arjun@confluent.io>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by wicknicks** [comment]
@rhauch @mageshn can you give this a first pass? thanks a lot!

File: core/src/main/scala/kafka/utils/Logging.scala
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/utils/Logging.scala:31]: Could we keep this log message?

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: Should we move this docstring to the interface itself?

File: core/src/main/scala/kafka/utils/Logging.scala
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/utils/Logging.scala]: nit: unused import `Log4jController `

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: nit: typo - `lo4j`. while you're at it, maybe "control **the** log levels" ?

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: This isn't a REST extension necessarily, right? It's also used by Kafka via JMX. I think mentioning `worker restarts` and `rest extension` might be confusing

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jControllerMBean.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jControllerMBean.java:24]: nit: What do you think about sprinkling a bit of docstrings in this interface while we're rewriting it? I like the original description of `An MBean that allows the user to dynamically alter log4j levels at runtime.`

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:86]: this _technically_ changes the public interface and would require a KIP if we're being pedantic about the process. I personally think we can go by without a KIP but we obviously need a committer to say what they think

**Review by stanislavkozlovski — COMMENTED** [review]
After testing this locally and comparing to the old version, I find that this now exposes all Kafka Java classes' log levels.
Previously, the Scala implementation would not let you get/set Java classes' log levels. 
e.g `kafka.server.FetchManager`

Considering that, I feel like this change definitely warrants a KIP and discussion around it

File: core/src/main/scala/kafka/utils/Logging.scala
Comment (mageshn) [reply core/src/main/scala/kafka/utils/Logging.scala:31]: Agreed. But it would be useful to add it in `LogLevelManager`

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (mageshn) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:86]: @stanislavkozlovski Did you intend to say that the whole MBean is a public interface?

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (mageshn) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:34]: @wicknicks would be useful to have some unit test for this class.

**Review by mageshn — COMMENTED** [review]
@wicknicks This will be a super useful capability. Also, have we considered upgrading the log4j version which helps provide dynamic log levels?

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:86]: I meant that the old implementation would return `Null log level.` when there wasn't a log level, whereas this would return the effective level. It's insignificant and that's why I mentioned that we probably don't need to be pedantic.

See my other comment about this PR exposing more classes though. I believe this needs a KIP

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (rodesai) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:44]: Is this the format required by the mbean stuff? It might be nice to return something more structured here.

**Review by rodesai — APPROVED** [review]
LGTM. Agree on the added doc-strings

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (rayokota) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: Can `LogManager.getRootLogger().getLevel()` be `null`?  With other loggers you return the effective level if that's the case.

**Review by rayokota — APPROVED** [review]
Just a minor question, otherwise LGTM.

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:44]: we can only return primitives in MBean (also was trying to preserve the previous API behavior). Look here, also for problems when using complex types in MBeans: https://www.oracle.com/technetwork/java/javase/tech/best-practices-jsp-136021.html#mozTocId931827

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: I couldn't find a way to initialize the root level to be null. Maybe that happens if log4j was not initialized properly. There is a check in `LogLevelManager` that will prevent the MBean from being registered if log4j classes are not in classpath.

File: core/src/main/scala/kafka/utils/Logging.scala
Comment (wicknicks) [reply core/src/main/scala/kafka/utils/Logging.scala:31]: this is has been added in `LogLevelManager`.

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:86]: posted a KIP here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-495%3A+Dynamically+Adjust+Log+Levels+in+Connect (feedback welcome!)

File: core/src/main/scala/kafka/utils/Logging.scala
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/utils/Logging.scala]: nit: `{}` are unneeded

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:34]: added in Log4jControllerTest below. thanks!

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jControllerMBean.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jControllerMBean.java:24]: added some docs here. let me know what you think.

**Comment by stanislavkozlovski** [comment]
Regarding my previous comment:
> After testing this locally and comparing to the old version, I find that this now exposes all Kafka Java classes' log levels.
Previously, the Scala implementation would not let you get/set Java classes' log levels.
e.g kafka.server.FetchManager

I tested again and found this to be incorrect. First of all, kafka.server.FetchManager is not a Java class, it's scala. We always seemed to have exposed the Java class loggers

File: build.gradle
Comment (stanislavkozlovski) [inline build.gradle:949]: Could we remove the `compileOnly libs.log4j` from the project('core') dependencies? (~line 680) I think we don't need it

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java]: Since this doesn't apply to Connect only, could we say something like "across application restarts"?  

Also, does `<p />` mean something like a paragraph or is it a typo?

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: Is there a reason to be using a `Hashtable` as opposed to a `HashMap`?

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java:82]: Could we add a test for the `Logger#effectiveLogLevel()` call?

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: nit: `Logger h` should work as well

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: nit: the signature for assertEquals is `assertEquals(Object expected, Object actual)` -- we want the expected value to be first, in this case "WARN"

Ditto for the test below

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: nit: Could we use more than one character for variable names? Test code typically gets read and edited as much as production code, so it would be good to hold it to the same quality standard

**Comment by wicknicks** [comment]
@stanislavkozlovski thanks for the test! Yeah, I don't think that should happen. I tried to retain the existing behavior as much as possible. One big difference is using effective log level and removing newLogger [1] from the old code (I don't think we should create loggers if they don't exist). The new code only allows you to set/get log levels of existing loggers.

[1] https://github.com/apache/kafka/pull/6069/files#diff-ce366937756affa54fe2b50feff24f23L49

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (stanislavkozlovski) [inline clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:60]: What are your thoughts regarding returning the same `"No such logger"` value? It might be more informative to JMX users

**Review by stanislavkozlovski — COMMENTED** [review]
I gave the PR another pass. Left a couple of small comments, mostly looks good

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (wicknicks) [reply clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: it is used so we can create an Enumeration<> below on line 46.

File: build.gradle
Comment (wicknicks) [reply build.gradle:949]: done. thanks!

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (wicknicks) [reply clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java:82]: done. thanks!

File: clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java
Comment (wicknicks) [reply clients/src/main/java/org/apache/kafka/common/logging/Log4jController.java:60]: I'm ok with either. Will leave this here for other reviewers to also take a look before making a final decision.

**Review by wicknicks — COMMENTED** [review]
@stanislavkozlovski thanks for your feedback, addressed your comments!

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [reply clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java]: Could we achieve the same thing with `Collections.enumeration(hashmap.keySet())`?

File: clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java
Comment (stanislavkozlovski) [inline clients/src/test/java/org/apache/kafka/common/logging/Log4jControllerTest.java:36]: The previous comment got resolved due to the force push.

> @stanislavkozlovski: Is there a reason to be using a Hashtable as opposed to a HashMap?

> @wicknicks: it is used so we can create an Enumeration<> below on line 46.

> @stanislavkozlovski: Could we achieve the same thing with Collections.enumeration(hashmap.keySet())?

**Comment by wicknicks** [comment]
closed in favor of https://github.com/apache/kafka/pull/7403

**Pull request closed without merging**
